File: callouts/python/extproc/tests/jwt_auth_test.py
# Copyright 2024 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import print_function

import datetime
import re

import jwt
from envoy.config.core.v3.base_pb2 import HeaderMap
from envoy.config.core.v3.base_pb2 import HeaderValue
from envoy.service.ext_proc.v3 import external_processor_pb2 as service_pb2
from envoy.service.ext_proc.v3 import external_processor_pb2_grpc as service_pb2_grpc

import grpc
import pytest

from extproc.example.jwt_auth.service_callout_example import (
    CalloutServerExample as CalloutServerTest)
from extproc.service.callout_tools import add_header_mutation
from extproc.tests.basic_grpc_test import (
    make_request,
    setup_server,
    get_insecure_channel,
    insecure_kwargs,
)


# Import the setup server test fixture.
_ = setup_server
_local_test_args = {"kwargs": insecure_kwargs, "test_class": CalloutServerTest}

@pytest.mark.parametrize('server', [_local_test_args], indirect=True)
def test_jwt_auth_rs256_failure(server: CalloutServerTest) -> None:
  with get_insecure_channel(server) as channel:
    stub = service_pb2_grpc.ExternalProcessorStub(channel)

    # Construct the HeaderMap
    header_map = HeaderMap()
    header_value = HeaderValue(key="Authorization", raw_value=b"")
    header_map.headers.extend([header_value])

    # Construct HttpHeaders with the HeaderMap
    request_headers = service_pb2.HttpHeaders(headers=header_map,
                                              end_of_stream=True)

    # Use request_headers in the request
    with pytest.raises(grpc.RpcError) as e:
      make_request(stub, request_headers=request_headers)
    assert e.value.code() == grpc.StatusCode.PERMISSION_DENIED

@pytest.mark.parametrize('server', [_local_test_args], indirect=True)
def test_jwt_auth_rs256_success(server: CalloutServerTest) -> None:
    with get_insecure_channel(server) as channel:
        stub = service_pb2_grpc.ExternalProcessorStub(channel)

        # Load the private key
        private_key_path = './extproc/ssl_creds/localhost.key'
        with open(private_key_path, 'r') as key_file:
            private_key = key_file.read()

        # Define the payload for the JWT
        payload = {
            "sub": "1234567890",
            "name": "John Doe",
            "admin": True,
            "iat": datetime.datetime.utcnow(),
            "exp": datetime.datetime.utcnow() + datetime.timedelta(hours=1)
        }

        # Generate the JWT token
        jwt_token = jwt.encode(payload, private_key, algorithm="RS256")

        # Authorization header value
        authorization_header_value = f"Bearer {jwt_token}"

        # Construct the HeaderMap
        header_map = HeaderMap()
        header_value = HeaderValue(key="Authorization", raw_value=bytes(authorization_header_value, 'utf-8'))
        header_map.headers.extend([header_value])

        # Construct HttpHeaders with the HeaderMap
        request_headers = service_pb2.HttpHeaders(headers=header_map, end_of_stream=True)

        # Construct the decoded items list from the payload
        decoded_items = [(f'decoded-{key}', str(value)) for key, value in payload.items() if key != 'exp' and key != 'iat']
        # Adding formatted 'iat' and 'exp' to match the test format
        decoded_items.extend([
            ('decoded-iat', str(int(payload['iat'].timestamp()))),
            ('decoded-exp', str(int(payload['exp'].timestamp())))
        ])

        value = make_request(stub, request_headers=request_headers)
        assert value.HasField('request_headers')
        # Instead of directly comparing the full response, check the presence and basic validation of decoded items
        assert 'header_mutation' in str(value)
        for key, expected_value in decoded_items:
          # Check presence of key
          assert key in str(value)
          # For 'iat' and 'exp', check if it matches the pattern since the value will be different
          if key in ['decoded-iat', 'decoded-exp']:
            pattern = rf'{key}"\s*raw_value:\s*"\d+"'
            assert re.search(pattern, str(value)), f"{key} does not match expected pattern"
          else:
            # For other keys, check the exact value
            pattern = rf'{key}"\s*raw_value:\s*"{expected_value}"'
            assert re.search(pattern, str(value)), f"{key} value {expected_value} not found"

File: callouts/go/extproc/tests/jwt_auth_test.go
// Copyright 2024 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package tests

import (
	"fmt"
	"github.com/dgrijalva/jwt-go"
	base "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
	extproc "github.com/envoyproxy/go-control-plane/envoy/service/ext_proc/v3"
	"github.com/stretchr/testify/assert"
	"io/ioutil"
	"regexp"
	"service-extensions-samples/extproc/examples/jwt_auth"
	"testing"
	"time"
)

func generateTestJWTToken(privateKey []byte, claims jwt.MapClaims) (string, error) {
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	privateKeyParsed, err := jwt.ParseRSAPrivateKeyFromPEM(privateKey)
	if err != nil {
		return "", err
	}
	return token.SignedString(privateKeyParsed)
}

func TestHandleRequestHeaders_ValidToken(t *testing.T) {
	// Load test keys
	privateKey, err := ioutil.ReadFile("../ssl_creds/localhost.key")
	assert.NoError(t, err, "failed to load private key")

	// Create test JWT token
	now := time.Now().Unix()
	claims := jwt.MapClaims{
		"sub":  "1234567890",
		"name": "John Doe",
		"iat":  now,
		"exp":  now + 3600,
	}
	tokenString, err := generateTestJWTToken(privateKey, claims)
	assert.NoError(t, err, "failed to generate test JWT token")

	headers := &extproc.HttpHeaders{
		Headers: &base.HeaderMap{
			Headers: []*base.HeaderValue{
				{
					Key:      "Authorization",
					RawValue: []byte("Bearer " + tokenString),
				},
			},
		},
	}

	// Create an instance of ExampleCalloutService with overridden public key path
	service := jwt_auth.NewExampleCalloutServiceWithKeyPath("../ssl_creds/publickey.pem")

	// Call the HandleRequestHeaders method
	response, err := service.HandleRequestHeaders(headers)

	// Assert that no error occurred
	assert.NoError(t, err)

	// Assert that the response is not nil
	assert.NotNil(t, response)

	// Prepare expected decoded items without iat and exp
	expectedItems := []struct{ Key, Value string }{
		{"decoded-sub", "1234567890"},
		{"decoded-name", "John Doe"},
	}

	// Validate the response headers
	requestHeaders := response.GetRequestHeaders().GetResponse()
	for _, item := range expectedItems {
		found := false
		for _, header := range requestHeaders.GetHeaderMutation().GetSetHeaders() {
			if header.GetHeader().GetKey() == item.Key && string(header.GetHeader().GetRawValue()) == item.Value {
				found = true
				break
			}
		}
		assert.True(t, found, fmt.Sprintf("header %s: %s not found in response", item.Key, item.Value))
	}

	// Validate iat and exp separately using regex patterns
	iatPattern := regexp.MustCompile(`^\d+(\.\d+)?(e[\+\-]?\d+)?$`)
	expPattern := regexp.MustCompile(`^\d+(\.\d+)?(e[\+\-]?\d+)?$`)

	iatFound := false
	expFound := false
	for _, header := range requestHeaders.GetHeaderMutation().GetSetHeaders() {
		key := header.GetHeader().GetKey()
		value := string(header.GetHeader().GetRawValue())
		if key == "decoded-iat" {
			assert.True(t, iatPattern.MatchString(value), fmt.Sprintf("decoded-iat value %s does not match expected pattern", value))
			iatFound = true
		}
		if key == "decoded-exp" {
			assert.True(t, expPattern.MatchString(value), fmt.Sprintf("decoded-exp value %s does not match expected pattern", value))
			expFound = true
		}
	}

	assert.True(t, iatFound, "header decoded-iat not found or mismatched in response")
	assert.True(t, expFound, "header decoded-exp not found or mismatched in response")
}

func TestHandleRequestHeaders_InvalidToken(t *testing.T) {

	headers := &extproc.HttpHeaders{
		Headers: &base.HeaderMap{
			Headers: []*base.HeaderValue{
				{
					Key:      "Authorization",
					RawValue: []byte("Bearer invalidtoken"),
				},
			},
		},
	}

	// Create an instance of ExampleCalloutService with test public key path
	service := jwt_auth.NewExampleCalloutServiceWithKeyPath("../ssl_creds/publickey.pem")

	service.PublicKey = []byte("invalidpublickey")

	// Call the HandleRequestHeaders method
	_, err := service.HandleRequestHeaders(headers)

	// Assert that an error occurred
	assert.Error(t, err)

	// Assert that the error is a permission denied error
	assert.Contains(t, err.Error(), "PermissionDenied")

	// Assert that the error message contains "Authorization token is invalid"
	assert.Contains(t, err.Error(), "Authorization token is invalid")
}
